### PR TITLE
[SCHEDULE]Improve bound deduce for loop partition (#743)

### PR DESCRIPTION
**Pull request by xqdan**
* add divided checking when deducing
* related testcase

### CODE REVIEW

File: src/arithmetic/bound_deducer.cc [hunk 1]
```diff
@@ -128,7 +128,14 @@ class BoundDeducer: public IRVisitor {
     }
 
     // always use relax bound
-    result = result / operand + (is_greater ? 1 : -1);
+    Expr div_result = result / operand;
```
Comment (tqchen) [inline src/arithmetic/bound_deducer.cc]: use instead can_prove(result % operand == 0)

File: src/arithmetic/bound_deducer.cc [hunk 2]
```diff
@@ -128,7 +128,14 @@ class BoundDeducer: public IRVisitor {
...
+    bool divided = can_prove(div_result * operand == result);
+    result = div_result;
+    if (is_greater) {
+      result += (divided && is_equal) ? 0 : 1;
```
Comment (tqchen) [inline src/arithmetic/bound_deducer.cc]: This piece of code is cryptic, need proper comment on why do we need repeatitive is_equal checks

File: src/arithmetic/bound_deducer.cc [hunk 3]
```diff
@@ -237,9 +244,9 @@ IntSet DeduceBound(Expr v, Expr e,
   if (!d.success) return IntSet::nothing();
   Expr min = Interval::neg_inf, max = Interval::pos_inf;
   if (d.is_greater) {
-    min = d.is_equal ? d.result : d.result + 1;
+    min = d.result;
   } else {
-    max = d.is_equal ? d.result : d.result - 1;
+    max = d.result;
```
Comment (tqchen) [inline src/arithmetic/bound_deducer.cc]: As the deduction get complicated for checking both cases if there is equality in the comparison. I would recommend move it away by first transforming the equation to canonical form like 

```(a <= b)``` and have a single point of logic for doing the deduction

**Comment by tqchen** [comment]
add @ZihengJiang to the reviewer

File: src/arithmetic/bound_deducer.cc (hunk 3 above)
Comment (tqchen) [reply src/arithmetic/bound_deducer.cc]: specifically, we want to canonicalize the result if the is_equal check is false, and move to the case when is_equal is true.

File: tests/python/unittest/test_arith_intset.py [hunk 4]
```diff
@@ -59,14 +59,42 @@ def test_check():
...
+        res1 = tvm.arith.DeduceBound(a, e0<=17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.max() * 4 + b_s.max()) <= 17)).value == 1
+      
+        res1 = tvm.arith.DeduceBound(a, e0>=17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.min() * 4 + b_s.min()) >= 17)).value == 1
+        
+
+    test(-1, 2)
+    test(10, 17)
+    test(1, 5)
```
Comment (tqchen) [inline tests/python/unittest/test_arith_intset.py]: need to add testcase when a is a bigger negative number than 1

File: src/arithmetic/bound_deducer.cc (hunk 3 above)
Comment (xqdan) [reply src/arithmetic/bound_deducer.cc]: Ok, the way lt->le, gt->ge is better indeed.

File: src/arithmetic/bound_deducer.cc [hunk 5]
```diff
@@ -128,13 +128,23 @@ class BoundDeducer: public IRVisitor {
...
+    result = result / operand;
+    // since system will round down when not divided
+    // eg. 2/4 -> 0; -2/4 -> -1
+    // no need fix for !is_greater:
+    // eg. a <= 2/4 -> a <= 0
+    // eg. a <= 0/4 -> a <= 0
+    // so just fix for not divided and is_greater
+    // eg. a >= 2/4 -> a >= 0 + 1
+    // eg. a >= 0/4 -> a >= 0
+    result += (is_greater && !divided) ? 1 : 0;
```
Comment (tqchen) [inline src/arithmetic/bound_deducer.cc]: Use 
```
if (is_greater && !divided) {
   result += 1;
}
```

File: tests/python/unittest/test_arith_intset.py [hunk 6]
```diff
@@ -59,14 +59,37 @@ def test_check():
...
+
+        res1 = tvm.arith.DeduceBound(a, e0<=17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.max() * 4 + b_s.max()) <= 17)).value == 1
+      
+        res1 = tvm.arith.DeduceBound(a, e0>=17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.min() * 4 + b_s.min()) >= 17)).value == 1
+       
+    test(0, 4)
+    test(1, 5)
+    test(2, 6)
```
Comment (tqchen) [inline tests/python/unittest/test_arith_intset.py]: for better coverage, we need to be able to give case when the co-efficient is negative

Comment (tqchen) [reply tests/python/unittest/test_arith_intset.py]: add a testcase with more complicated nest

Comment (xqdan) [reply tests/python/unittest/test_arith_intset.py]: do you mean e0 = b + a*(-4)?

File: tests/python/unittest/test_arith_intset.py [hunk 7]
```diff
@@ -59,14 +64,57 @@ def test_check():
...
+        res1 = tvm.arith.DeduceBound(a, e0>17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.min() * 4 + 3 + 2*b_s.min()) > 17)).value == 1
+
+        res1 = tvm.arith.DeduceBound(a, e0<=17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.max() * 4 + 3 + 2*b_s.max()) <= 17)).value == 1
+      
+        res1 = tvm.arith.DeduceBound(a, e0>=17, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.min() * 4 + 3 + 2*b_s.min()) >= 17)).value == 1
+   
+    def test_neg(a1, a2):
```
Comment (tqchen) [inline tests/python/unittest/test_arith_intset.py]: combine test_pos and test_neg, pass the coefficient as argument.

File: tests/python/unittest/test_arith_intset.py [hunk 8]
```diff
@@ -59,14 +64,57 @@ def test_check():
...
+
+        res1 = tvm.arith.DeduceBound(a, e0>21, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.max() * (-4) + 3 + (-2)*b_s.min()) > 21)).value == 1
+
+        res1 = tvm.arith.DeduceBound(a, e0<=21, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.min() * (-4) + 3 + (-2)*b_s.max()) <= 21)).value == 1
+
+        res1 = tvm.arith.DeduceBound(a, e0>=21, {b: b_s}, {b: b_s})
+        assert (tvm.ir_pass.Simplify((res1.max() * (-4) + 3 + (-2)*b_s.min()) >= 21)).value == 1
+
```
Comment (tqchen) [inline tests/python/unittest/test_arith_intset.py]: add a testcase with more level of nesting, like
```
(x * b + a* 2) * 4
```

Comment (xqdan) [reply tests/python/unittest/test_arith_intset.py]: a = tvm.var('a')
        b = tvm.var('b')
        c = tvm.var('c')
        b_s = tvm.arith.intset_interval(a1, a2)
        c_s = tvm.arith.intset_interval(10, 20)
        e0 = (b*c + a* coff) * 4

        res1 = tvm.arith.DeduceBound(a, e0<63, {b: b_s}, {b: b_s})

res1.max() = ((((63 - 1)/4) - select((c >= 0), (4*c), (0*c)))/4) when coff > 0

I guess there is work to do

**Comment by xqdan** [comment]
```
        a = tvm.var('a')
        b = tvm.var('b')
        c = tvm.var('c')
        b_s = tvm.arith.intset_interval(a1, a2)
        c_s = tvm.arith.intset_interval(10, 20)
        e0 = (b*c + a* coff) * 4

        res1 = tvm.arith.DeduceBound(a, e0<63, {b: b_s, c: c_s}, {b: b_s, c: c_s})
```

res1.max() = neg_inf

[12:02:36] src/arithmetic/int_set.cc:256: Return Everything in CombineInterval Mul

```
inline IntSet CombineInterval<Mul>(Interval a, Interval b) {
  if (a.is_single_point() && b.is_single_point()) {
    return IntSet::single_point(ComputeExpr<Mul>(a.min, b.min));
  }
  if (a.is_single_point() && !b.is_single_point()) {
    std::swap(a, b);
  }
  if (b.is_single_point()) {
  }
  LOG(WARNING) << "Return Everything in CombineInterval Mul";
```
no !a.is_single_point()&&!b.is_single_point() support here
I guess there is work to do

**Comment by tqchen** [comment]
interval combination is a more rare case that we do not yet need to support, modify the test-case for now and restrict it to contain one interval without multiply

**Comment by tqchen** [comment]
specifically, set c to be a constant number in the above testcase

**Pull request merged**